File: frontend/services/reunionService.ts
import { getAuthToken } from "./authService";

const API_BASE = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080";

export interface ReunionMediaDto {
  id: string;
  url: string;
  mediaType: string;
  caption?: string;
  creatorId: string;
  creatorName: string;
}

export interface ReunionCommentDto {
  id: string;
  userId: string;
  userFullName: string;
  userProfilePicture?: string;
  text: string;
  createdAt: string;
}

export interface ReunionCollectionDto {
  id: string;
  title: string;
  description?: string;
  date: string;
  location?: string;
  photos: ReunionMediaDto[];
  comments: ReunionCommentDto[];
  attendeesCount: number;
  attending: boolean;
}

async function authFetch(url: string, options: RequestInit = {}) {
  const token = getAuthToken();
  if (!token) throw new Error("No authentication token found.");
  const res = await fetch(url, {
    ...options,
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json", ...options.headers },
  });
  if (!res.ok) throw new Error(await res.text() || "Request failed");
  return res;
}

export async function getReunions(): Promise<ReunionCollectionDto[]> {
  const res = await authFetch(`${API_BASE}/api/reunions`);
  return res.json();
}

export async function createReunion(title: string, description: string, date: string, location: string): Promise<ReunionCollectionDto> {
  const qs = new URLSearchParams({ title, description, date, location }).toString();
  const res = await authFetch(`${API_BASE}/api/reunions?${qs}`, { method: "POST" });
  return res.json();
}

export async function rsvpReunion(id: string): Promise<ReunionCollectionDto> {
  const res = await authFetch(`${API_BASE}/api/reunions/${id}/rsvp`, { method: "POST" });
  return res.json();
}

export async function withdrawReunion(id: string): Promise<ReunionCollectionDto> {
  const res = await authFetch(`${API_BASE}/api/reunions/${id}/withdraw`, { method: "POST" });
  return res.json();
}

export async function addReunionMedia(id: string, url: string, mediaType: string, caption?: string): Promise<ReunionMediaDto> {
  const qs = new URLSearchParams({ url, mediaType, ...(caption ? { caption } : {}) }).toString();
  const res = await authFetch(`${API_BASE}/api/reunions/${id}/media?${qs}`, { method: "POST" });
  return res.json();
}

export async function addReunionComment(id: string, text: string): Promise<ReunionCommentDto> {
  const res = await authFetch(`${API_BASE}/api/reunions/${id}/comment?text=${encodeURIComponent(text)}`, { method: "POST" });
  return res.json();
}

File: backend/src/main/java/com/alumnihub/service/ReunionService.java
package com.alumnihub.service;

import com.alumnihub.dto.ReunionCollectionDto;
import com.alumnihub.dto.ReunionCommentDto;
import com.alumnihub.dto.ReunionMediaDto;
import com.alumnihub.entity.*;
import com.alumnihub.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ReunionService {

    private final ReunionCollectionRepository reunionCollectionRepository;
    private final ReunionMediaRepository reunionMediaRepository;
    private final ReunionCommentRepository reunionCommentRepository;
    private final UserRepository userRepository;

    @Transactional
    public ReunionCollectionDto createReunion(String title, String description, LocalDateTime date, String location) {
        ReunionCollection reunion = ReunionCollection.builder()
                .title(title)
                .description(description)
                .date(date)
                .location(location)
                .build();

        ReunionCollection saved = reunionCollectionRepository.save(reunion);
        return convertToDto(saved, null);
    }

    public List<ReunionCollectionDto> getReunions(String userEmail) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        return reunionCollectionRepository.findAll().stream()
                .map(r -> convertToDto(r, user))
                .collect(Collectors.toList());
    }

    @Transactional
    public ReunionCollectionDto rsvpReunion(String userEmail, UUID reunionId) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        ReunionCollection r = reunionCollectionRepository.findById(reunionId)
                .orElseThrow(() -> new IllegalArgumentException("Reunion not found: " + reunionId));

        r.getAttendees().add(user);
        return convertToDto(reunionCollectionRepository.save(r), user);
    }

    @Transactional
    public ReunionCollectionDto withdrawReunion(String userEmail, UUID reunionId) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        ReunionCollection r = reunionCollectionRepository.findById(reunionId)
                .orElseThrow(() -> new IllegalArgumentException("Reunion not found: " + reunionId));

        r.getAttendees().remove(user);
        return convertToDto(reunionCollectionRepository.save(r), user);
    }

    @Transactional
    public ReunionMediaDto addMedia(String userEmail, UUID reunionId, String url, String mediaType, String caption) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        ReunionCollection r = reunionCollectionRepository.findById(reunionId)
                .orElseThrow(() -> new IllegalArgumentException("Reunion not found: " + reunionId));

        ReunionMedia media = ReunionMedia.builder()
                .reunion(r)
                .url(url)
                .mediaType(mediaType)
                .caption(caption)
                .creator(user)
                .build();

        ReunionMedia saved = reunionMediaRepository.save(media);
        return convertMediaToDto(saved);
    }

    @Transactional
    public ReunionCommentDto addComment(String userEmail, UUID reunionId, String text) {
        User user = userRepository.findByEmail(userEmail)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userEmail));

        ReunionCollection r = reunionCollectionRepository.findById(reunionId)
                .orElseThrow(() -> new IllegalArgumentException("Reunion not found: " + reunionId));

        ReunionComment comment = ReunionComment.builder()
                .reunion(r)
                .user(user)
                .text(text)
                .build();

        ReunionComment saved = reunionCommentRepository.save(comment);
        return convertCommentToDto(saved);
    }

    private ReunionCollectionDto convertToDto(ReunionCollection r, User user) {
        return ReunionCollectionDto.builder()
                .id(r.getId())
                .title(r.getTitle())
                .description(r.getDescription())
                .date(r.getDate())
                .location(r.getLocation())
                .photos(r.getPhotos().stream().map(this::convertMediaToDto).collect(Collectors.toList()))
                .comments(r.getComments().stream().map(this::convertCommentToDto).collect(Collectors.toList()))
                .attendeesCount(r.getAttendees().size())
                .attending(user != null && r.getAttendees().contains(user))
                .build();
    }

    private ReunionMediaDto convertMediaToDto(ReunionMedia rm) {
        return ReunionMediaDto.builder()
                .id(rm.getId())
                .url(rm.getUrl())
                .mediaType(rm.getMediaType())
                .caption(rm.getCaption())
                .creatorId(rm.getCreator().getId())
                .creatorName(rm.getCreator().getFullName())
                .build();
    }

    private ReunionCommentDto convertCommentToDto(ReunionComment rc) {
        return ReunionCommentDto.builder()
                .id(rc.getId())
                .userId(rc.getUser().getId())
                .userFullName(rc.getUser().getFullName())
                .userProfilePicture(rc.getUser().getProfilePictureUrl())
                .text(rc.getText())
                .createdAt(rc.getCreatedAt())
                .build();
    }
}
